### PR TITLE
[Feature][Stream Load] Add "show stream load" to show stream load record

### PR DESCRIPTION
**Pull request by weizuo93**
## Proposed changes

Add the sql of `show stream load` to show the result of recent stream load jobs. 

img_58a449aa-9d4b-4bbc-85a8-cbd64fc7659l

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [] I have added tests that prove my fix is effective or that my feature works
- [] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

### CODE REVIEW

File: be/src/http/action/stream_load.cpp [hunk 1]
```diff
@@ -132,9 +134,19 @@ void StreamLoadAction::handle(HttpRequest* req) {
     str = str + '\n';
     HttpChannel::send_reply(req, str);
 
+    auto stream_load_record = StorageEngine::instance()->get_stream_load_record();
+    if (stream_load_record != nullptr) {
+        auto st = stream_load_record->put(ToStringFromUnixMicros(ctx->start_micros + ctx->load_cost_micros) + "_" + ctx->label, str);
+        if (st.ok()) {
+            LOG(INFO) << "put stream_load_record rocksdb successfully. label: " << ctx->label;
```
Comment (acelyc111) [inline be/src/http/action/stream_load.cpp:141]: Better to log the whole key, then we can check it in rocksdb.

File: be/src/http/action/stream_load.cpp [hunk 2]
```diff
@@ -132,9 +134,19 @@ void StreamLoadAction::handle(HttpRequest* req) {
...
+        }
+    } else {
+        LOG(WARNING) << "stream_load_record is null.";
```
Comment (acelyc111) [inline be/src/http/action/stream_load.cpp:144]:
```suggestion
        LOG(WARNING) << "put stream_load_record rocksdb failed, status = " << st.ToString();
```
And log the key.

File: be/src/http/action/stream_load.cpp [hunk 3]
```diff
@@ -200,6 +212,16 @@ int StreamLoadAction::on_header(HttpRequest* req) {
...
         HttpChannel::send_reply(req, str);
+
+        auto stream_load_record = StorageEngine::instance()->get_stream_load_record();
+        if (stream_load_record != nullptr) {
+            auto st = stream_load_record->put(ToStringFromUnixMicros(ctx->start_micros + ctx->load_cost_micros) + "_" + ctx->label, str);
+            if (st.ok()) {
+                LOG(INFO) << "put stream_load_record rocksdb successfully. label: " << ctx->label;
+            }
+        } else {
+            LOG(WARNING) << "stream_load_record is null.";
```
Comment (acelyc111) [inline be/src/http/action/stream_load.cpp:223]: Same as above.

File: be/src/olap/olap_server.cpp [hunk 4]
```diff
@@ -118,6 +118,13 @@ Status StorageEngine::start_bg_threads() {
         LOG(INFO) << "path scan/gc threads started. number:" << get_stores().size();
     }
 
+    // stream load record clean thread
+    RETURN_IF_ERROR(Thread::create(
```
Comment (acelyc111) [inline be/src/olap/olap_server.cpp:122]: No need to create a thread to gc expired data, you can create a Compaction-Filter on this rocksdb instance.
https://github.com/facebook/rocksdb/wiki/Compaction-Filter

File: be/src/runtime/stream_load/stream_load_record.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        LOG(WARNING) << "rocks db put key:" << key << " failed, reason:" << s.ToString();
+        return Status::InternalError("Stream load record rocksdb put failed");
+    }
+    return Status::OK();
+}
+
+Status StreamLoadRecord::get_batch(const std::string& start, const int batch_size, std::map<std::string, std::string> &stream_load_records) {
+    rocksdb::ColumnFamilyHandle* handle = _handles[1];
+    std::unique_ptr<rocksdb::Iterator> it(_db->NewIterator(rocksdb::ReadOptions(), handle));
+    if (start == "") {
```
Comment (acelyc111) [inline be/src/runtime/stream_load/stream_load_record.cpp:86]:
```suggestion
    if (start.empty()) {
```

File: be/src/runtime/stream_load/stream_load_record.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        it->SeekToFirst();
+    } else {
+        it->Seek(start);
+        rocksdb::Status status = it->status();
+        if (!status.ok()) {
+            it->SeekToFirst();
```
Comment (acelyc111) [inline be/src/runtime/stream_load/stream_load_record.cpp:92]: why seek to first in this case?

File: be/src/runtime/stream_load/stream_load_record.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+    rocksdb::Status status = it->status();
+    if (!status.ok()) {
+        LOG(WARNING) << "rocksdb seek failed. reason:" << status.ToString();
+        return Status::InternalError("Stream load record rocksdb seek failed");
+    }
+    int num = 0;
+    for (it->Next(); it->Valid(); it->Next()) {
```
Comment (acelyc111) [inline be/src/runtime/stream_load/stream_load_record.cpp:101]: Have you skipped the first k-v? should it be 
```
for (; it->Valid(); it->Next()) {
```
?

File: be/src/runtime/stream_load/stream_load_record.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    Status clean_expired_stream_load_record();
+
+private:
+    std::string _root_path;
+    rocksdb::DB* _db;
+    std::vector<rocksdb::ColumnFamilyHandle*> _handles;
+
+    enum ColumnFamilyIndex {
+        DEFAULT_COLUMN_FAMILY_INDEX = 0,
```
Comment (acelyc111) [inline be/src/runtime/stream_load/stream_load_record.h:49]: where does the default column family used for?

File: be/src/olap/storage_engine.cpp [hunk 9]
```diff
@@ -225,6 +226,35 @@ Status StorageEngine::_init_store_map() {
...
+    if (!st.ok()) {
+        LOG(WARNING) << "status=" << st.to_string();
+    }
+
+    return Status::OK();
+}
+
+Status StorageEngine::_init_stream_load_record() {
+    std::string stream_load_record_path;
+    if (_store_map.size() > 0) {
```
Comment (acelyc111) [inline be/src/olap/storage_engine.cpp:240]: Is it valid when _store_map is empty? Would it be better to explicit define the path in config file?

File: be/src/runtime/stream_load/stream_load_context.cpp [hunk 10]
```diff
@@ -89,6 +103,91 @@ std::string StreamLoadContext::to_json() const {
...
+void StreamLoadContext::parse_stream_load_record(std::string stream_load_record, TStreamLoadRecord& stream_load_item) {
+
+    rapidjson::Document document;
+    if (document.Parse(stream_load_record.data()).HasParseError()) {
+        return;
+    }
+
+    if (document.HasMember("Label")) {
+        const rapidjson::Value& label = document["Label"];
+        stream_load_item.__set_label(label.GetString());
```
Comment (acelyc111) [inline be/src/runtime/stream_load/stream_load_context.cpp:115]:
```suggestion
        stream_load_item.__set_label(document["Label"].GetString());
```

**Comment by kangkaisen** [comment]
@weizuo93 Hi, Thanks for your contribution.
I don't understand why we need to store the stream load meta both in FE and BE.  why we couldn't get stream load info from FE meta?

**Comment by weizuo93** [comment]
> @weizuo93 Hi, Thanks for your contribution.
> I don't understand why we need to store the stream load meta both in FE and BE. why we couldn't get stream load info from FE meta?
Our goal is to store the stream load meta in FE. The reason why the stream load meta is stored on BE first is to send the meta to FE in bulk to avoid calling RPC every time the stream load finished. After a certain period of time, stream load meta will be removed from BE.

**Comment by weizuo93** [comment]
> @weizuo93 Hi, Thanks for your contribution.
> I don't understand why we need to store the stream load meta both in FE and BE. why we couldn't get stream load info from FE meta?

Our goal is to store the stream load meta in FE. The reason why the stream load meta is stored on BE first is to send the meta to FE in bulk to avoid calling RPC every time the stream load finished. After a certain period of time, stream load meta will be removed from BE.

**Comment by kangkaisen** [comment]
@weizuo93 
Hi, I mean In `LoadManager`,  `dbIdToLabelToLoadJobs` already has the stream load join info, why we still need to store stream load job meta in BE?  in other words, I think we don't need to store stream load job meta in BE.

**Comment by weizuo93** [comment]
> join

@kangkaisen As far as I know, there is no info about stream load in `LoadManager`and `dbIdToLabelToLoadJobs`. `LoadManager` only manages bulk load.

**Comment by weizuo93** [comment]
> @weizuo93
> Hi, I mean In `LoadManager`, `dbIdToLabelToLoadJobs` already has the stream load join info, why we still need to store stream load job meta in BE? in other words, I think we don't need to store stream load job meta in BE.

@kangkaisen As far as I know, there is no info about stream load in LoadManagerand dbIdToLabelToLoadJobs. LoadManager only manages bulk load.

**Comment by weizuo93** [comment]
> @weizuo93 You are right. I am wrong. This PR could go ahead. (But store some load job meta in FE and store some load job meta in BE, which seems not very good)

@kangkaisen 
Thanks for your comment. There is code overlap between this PR and PR #5452(https://github.com/apache/incubator-doris/pull/5452). After discussion with Mingyu Chen, we first push forward PR #5452. About storing stream load meta in BE, it's a tradeoff. If BE reports the stream load result to FE  when each stream load job finished, it may result in too much stress for FE RPC. So the stream load results are temporarily stored on BE and then sent to FE by batch. After finished data sending, stream load meta will be removed from BE soon. Do you have any better idea?

**Comment by kangkaisen** [comment]
@weizuo93 Ok, Thanks for your explain. which is OK for me

**Comment by yangzhg** [comment]
`show stream load` is better, like `show routine load`

**Comment by weizuo93** [comment]
> `show stream load` is better, like `show routine load`

OK, thank you.

**Review by yangzhg — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/common/Config.java [hunk 11]
```diff
@@ -605,6 +605,12 @@
     @ConfField(mutable = true, masterOnly = true)
     public static int max_stream_load_record_size = 5000;
 
+    /**
+     * Whether to close show stream load and clear stream load records in memory.
+     */
+    @ConfField(mutable = true, masterOnly = true)
+    public static boolean close_show_stream_load = false;
```
Comment (yangzhg) [inline fe/fe-core/src/main/java/org/apache/doris/common/Config.java:612]: `disable_show_stream_load` is better

Comment (weizuo93) [reply fe/fe-core/src/main/java/org/apache/doris/common/Config.java]: > `disable_show_stream_load` is better

OK, I have modified it.

**Review by yangzhg — APPROVED** [review]
LGTM

**Pull request merged**